File: sim/world.ts
import { Land } from './land'
import { Barrier } from './barrier'
import { Food } from './food'
import { Cell } from './cell'
import { Plant } from './plant'
import { Spore } from './spore'
import { Entity } from './entity'
import { WeatherState, WorldState } from './types'
import { Vector } from './vector'

export interface WorldConfig {
  width: number
  height: number
  numColumns: number
  numRows: number
  deltaTime?: number
}

interface WindState {
  x: number; y: number; width: number; height: number
  dirX: number; dirY: number
  magnitude: number
  maxMagnitude: number
}

interface RainState {
  x: number; y: number; width: number; height: number
  magnitude: number
  maxMagnitude: number
}

interface SunState {
  x: number; y: number; width: number; height: number
  magnitude: number
}

const MAX_TEMPERATURE = 100

function isOverlapping(
  ax: number, ay: number, aw: number, ah: number,
  bx: number, by: number, bw: number, bh: number,
): boolean {
  const p = ax <= bx && bx <= ax + aw
  const q = ay <= by + bh && by + bh <= ay + ah
  const r = ay <= by && by <= ay + ah
  const s = ax <= bx + bw && bx + bw <= ax + aw
  return (q || r) && (p || s)
}

function calcWind(time: number, worldWidth: number, worldHeight: number, wind: WindState): void {
  const interval = 60, halfInterval = 30
  const approxLoc = Math.trunc(time * 10) % (interval * 10)
  const hshift = approxLoc - halfInterval * 10
  const prop = 1 - Math.abs(hshift) / (halfInterval * 10)

  wind.x = (worldWidth - wind.width) * prop
  wind.y = (Math.trunc(time / halfInterval) % (Math.trunc(worldHeight / wind.height) + 1)) * wind.height
  wind.magnitude = wind.maxMagnitude * prop
  wind.dirX = 1; wind.dirY = 0
}

function calcRain(time: number, worldWidth: number, worldHeight: number, rain: RainState): void {
  const interval = 60, halfInterval = 30
  const approxLoc = Math.trunc(time * 10) % (interval * 10)
  const hshift = approxLoc - halfInterval * 10
  const prop = 1 - Math.abs(hshift) / (halfInterval * 10)

  rain.x = (Math.trunc(time / halfInterval) % (Math.trunc(worldWidth / rain.width) + 1)) * rain.width
  rain.y = (worldHeight - rain.height) * prop
  rain.magnitude = rain.maxMagnitude * prop
}

function calcSun(time: number, worldWidth: number, sun: SunState): void {
  const interval = 240
  const sunWidth = 2 * worldWidth
  const dxdt = (worldWidth + sunWidth) / interval
  sun.x = worldWidth - (Math.trunc(time) % interval) * dxdt
  sun.y = 0
  sun.width = sunWidth
  sun.magnitude = 80
}

export class World {
  readonly width: number
  readonly height: number
  readonly numColumns: number
  readonly numRows: number
  readonly landWidth: number
  readonly landHeight: number
  readonly deltaTime: number

  worldTime: number

  lands: Land[]
  landById: Map<number, Land>

  entities: Map<string, Entity>
  barriers: Barrier[]

  wind: WindState
  rain: RainState
  sun: SunState

  constructor(config: WorldConfig) {
    this.width = config.width
    this.height = config.height
    this.numColumns = config.numColumns
    this.numRows = config.numRows
    this.deltaTime = config.deltaTime ?? 1 / 30

    this.worldTime = 0
    this.entities = new Map()
    this.barriers = []
    this.landById = new Map()

    const colRem = this.width % this.numColumns
    const rowRem = this.height % this.numRows
    const plotW = Math.floor((this.width - colRem) / this.numColumns)
    const plotH = Math.floor((this.height - rowRem) / this.numRows)
    this.landWidth = plotW
    this.landHeight = plotH

    this.lands = []
    let landId = 0
    for (let col = 0; col < this.numColumns; col++) {
      for (let row = 0; row < this.numRows; row++) {
        const w = plotW + (col === this.numColumns - 1 ? colRem : 0)
        const h = plotH + (row === this.numRows - 1 ? rowRem : 0)
        const land = new Land(landId++, col, row, col * plotW, row * plotH, w, h)
        land.nutrients = 100
        land.temperature = 50
        land.moisture = 50
        land.nutRegen = 5.5
        land.midTemp = row % 2 === 0 ? 30 : 50
        land.tempRadius = col % 2 === 0 ? 30 : 10
        land.updateFertility()
        this.lands.push(land)
        this.landById.set(land.id, land)
      }
    }

    this.wind = { x: 0, y: 0, width: 200, height: 200, dirX: 1, dirY: 0, magnitude: 0, maxMagnitude: 0 }
    this.rain = { x: 0, y: 0, width: 200, height: 200, magnitude: 0, maxMagnitude: 100 }
    this.sun = { x: 0, y: 0, width: 2 * this.width, height: this.height, magnitude: 0 }
  }

  getLand(px: number, py: number): Land | null {
    if (px < 0 || px >= this.width || py < 0 || py >= this.height) return null
    let col = Math.floor(px / this.landWidth)
    let row = Math.floor(py / this.landHeight)
    if (col >= this.numColumns) col = this.numColumns - 1
    if (row >= this.numRows) row = this.numRows - 1
    const idx = col * this.numRows + row
    return this.lands[idx] ?? null
  }

  getLandById(id: number | null): Land | null {
    if (id === null) return null
    return this.landById.get(id) ?? null
  }

  getEntityById(id: string | null): Entity | null {
    if (!id) return null
    return this.entities.get(id) ?? null
  }

  // ── Spawn methods ─────────────────────────────────────────────────────────

  spawnFood(x: number, y: number, attach = false): Food {
    const food = new Food(x, y)
    if (attach) food.attachToPoint(x, y, 0, 1000, 25)
    this.entities.set(food.id, food)
    return food
  }

  spawnCell(x: number, y: number): Cell {
    const cell = new Cell(x, y, this.width, this.height)
    this.entities.set(cell.id, cell)
    return cell
  }

  spawnPlant(x: number, y: number, dirX: number, dirY: number, segLength: number): Plant {
    const plant = new Plant(x, y, dirX, dirY, segLength)
    plant.attachToPoint(x, y, 0, 1000, 1)
    const land = this.getLand(x, y)
    if (land) plant.landId = land.id
    this.entities.set(plant.id, plant)
    return plant
  }

  spawnSpore(x: number, y: number, dirX: number, dirY: number, initialSpeed: number): Spore {
    const dir = new Vector(dirX, dirY, 0)
    const mag = dir.magnitude()
    if (Math.round(mag * 10) > 10 && mag !== 0) return this.spawnSpore(x, y, 0, 0, 0)

    const spore = new Spore(x, y)
    spore.growthDir = mag > 0 ? dir.normalize() : new Vector(1, 1, 0).normalize()
    if (mag > 0) {
      spore.vx = spore.growthDir.x * initialSpeed
      spore.vy = spore.growthDir.y * initialSpeed
    }
    this.entities.set(spore.id, spore)
    return spore
  }

  spawnLinearBarrier(originX: number, originY: number, c0: number, c1: number, xMin: number, xMax: number): Barrier {
    const b = new Barrier([c0, c1], 1, xMin, xMax, 5, 20, originX, originY)
    this.entities.set(b.id, b)
    this.barriers.push(b)
    return b
  }

  spawnParabolicBarrier(originX: number, originY: number, c0: number, c1: number, c2: number, xMin: number, xMax: number): Barrier {
    const b = new Barrier([c0, c1, c2], 2, xMin, xMax, 5, 20, originX, originY)
    this.entities.set(b.id, b)
    this.barriers.push(b)
    return b
  }

  spawnCubicBarrier(originX: number, originY: number, c0: number, c1: number, c2: number, c3: number, xMin: number, xMax: number): Barrier {
    const b = new Barrier([c0, c1, c2, c3], 3, xMin, xMax, 5, 20, originX, originY)
    this.entities.set(b.id, b)
    this.barriers.push(b)
    return b
  }

  // ── Step ──────────────────────────────────────────────────────────────────

  step(dt?: number): void {
    const stepTime = dt ?? this.deltaTime
    const finalTime = this.worldTime + stepTime

    while (this.worldTime < finalTime) {
      this.worldTime += this.deltaTime
      this.simulateEnvironment()
      this.simulateEntities()
    }
  }

  private simulateEnvironment(): void {
    calcWind(this.worldTime, this.width, this.height, this.wind)
    calcRain(this.worldTime, this.width, this.height, this.rain)
    calcSun(this.worldTime, this.width, this.sun)
    this.simulateLand()
  }

  private simulateLand(): void {
    for (const land of this.lands) {
      land.updateTemperature(this.worldTime)

      // Rain moisture
      if (isOverlapping(this.rain.x, this.rain.y, this.rain.width, this.rain.height, land.x, land.y, land.width, land.height)) {
        land.changeMoisture(this.rain.magnitude)
      }

      // Evaporation
      if (land.temperature > 70) {
        land.changeMoisture(-0.005 * (land.temperature / MAX_TEMPERATURE))
      }
      if (land.temperature > 95) {
        land.changeMoisture(-1 * (land.temperature / MAX_TEMPERATURE))
      }

      // Nutrients regen
      land.changeNutrients(land.nutRegen * this.deltaTime)

      // Sunlight
      if (isOverlapping(this.sun.x, this.sun.y, this.sun.width, this.sun.height, land.x, land.y, land.width, land.height)) {
        land.setSunlight(this.sun.magnitude)
      } else {
        land.setSunlight(50)
      }

      land.updateFertility()
    }
  }

  private simulateEntities(): void {
    const toDelete: string[] = []

    for (const [id, entity] of this.entities) {
      if (!entity.alive) {
        toDelete.push(id)
        continue
      }

      entity.accelerate(this.deltaTime)
      entity.viewMove(this.deltaTime)

      // Barrier collision
      for (const bar of this.barriers) {
        if (bar.alive && bar.isCollision(entity.x, entity.y)) {
          entity.collide(50000, 0, 0)
          if (entity.type === 'cell') (entity as Cell).feelBarrier()
          break
        }
      }

      // World boundary collision
      if (entity.x < 0 || entity.x >= this.width || entity.y < 0 || entity.y >= this.height) {
        entity.collide(300, 0, 0)
        // Clamp position inside bounds
        entity.x = Math.max(0, Math.min(this.width - 1, entity.x))
        entity.y = Math.max(0, Math.min(this.height - 1, entity.y))
      }

      // Per-type simulation
      if (entity.type === 'cell') {
        const cell = entity as Cell
        if (cell.x >= this.width || cell.x <= 0 || cell.y >= this.height || cell.y <= 0) {
          cell.feelBarrier()
        }
        cell.see(this)
        cell.simulate(this)

      } else if (entity.type === 'plant') {
        (entity as Plant).grow(this)

      } else if (entity.type === 'spore') {
        const spore = entity as Spore
        // Apply wind
        if (spore.x >= this.wind.x && spore.x <= this.wind.x + this.wind.width &&
            spore.y >= this.wind.y && spore.y <= this.wind.y + this.wind.height) {
          spore.applyForce(this.wind.dirX * this.wind.magnitude, this.wind.dirY * this.wind.magnitude, 0)
        }
        spore.lifeTime += this.deltaTime
        spore.simulate(this)

      } else if (entity.type === 'food') {
        const food = entity as Food
        if (food.x >= this.wind.x && food.x <= this.wind.x + this.wind.width &&
            food.y >= this.wind.y && food.y <= this.wind.y + this.wind.height) {
          food.applyForce(this.wind.dirX * this.wind.magnitude, this.wind.dirY * this.wind.magnitude, 0)
        }
      }
    }

    for (const id of toDelete) {
      this.entities.delete(id)
    }
  }

  // ── State export ──────────────────────────────────────────────────────────

  getState(): WorldState {
    const weather: WeatherState = {
      wind: { x: this.wind.x, y: this.wind.y, width: this.wind.width, height: this.wind.height, dirX: this.wind.dirX, dirY: this.wind.dirY, magnitude: this.wind.magnitude },
      rain: { x: this.rain.x, y: this.rain.y, width: this.rain.width, height: this.rain.height, magnitude: this.rain.magnitude },
      sun:  { x: this.sun.x,  y: this.sun.y,  width: this.sun.width,  height: this.sun.height,  magnitude: this.sun.magnitude },
    }

    return {
      time: this.worldTime,
      width: this.width,
      height: this.height,
      entities: Array.from(this.entities.values()).map(e => e.toJSON()),
      lands: this.lands.map(l => l.toJSON()),
      weather,
    }
  }
}

File: game_cpp/src/world.cpp
#include "world.h"

world::world(int cols, int rows, int width, int height, float _delta_time)
{
	num_columns = cols;
	num_rows = rows;
	world_width = width;
	world_height = height;


	SetRect(&brc, 0, 0, world_width, world_height);


	world_time = 0;
	delta_time = _delta_time;

	//Initialize land_plots.

	int column_rem = world_width % num_columns;
	int row_rem = world_height % num_rows;

	int plot_width = (world_width - column_rem) / num_columns;
	int plot_height = (world_height - row_rem) / num_rows;

	land_width = plot_width;
	land_height = plot_height;

	for (int x = 0; x < num_columns; x++)
	{
		for (int y = 0; y < num_rows; y++)
		{
			//(x, y) = top_left of current land plot

			int w = plot_width;
			int h = plot_height;

			POINT pt;
			pt.x = x * w;
			pt.y = y * h;


			if (x == num_columns - 1)
			{
				//On the last column. If there's a remainder, extend width to end.
				w += column_rem;
			}
			if (y == num_rows - 1)
			{
				//On the last row. If there's a remainder, extend height to end.
				h += row_rem;
			}


			land temp(pt, w, h);
			temp.nutrient_level = 100;
			temp.temperature = 50;
			temp.moisture_level = 50;			

			temp.nut_regen = 5.5;
			temp.mid_temp = 50;
			temp.temp_radius = 10;

			if (x % 2 == 0)
			{
				temp.temp_radius = 30;
			}
			if (y % 2 == 0)
			{
				temp.mid_temp = 30;
			}

			temp.calculate_fertility();

			land_plots.push_back(temp);

		}
	}





	//Initialize nature boxes

	//Sun
	SUN = new sun(world_width, world_height);






}


void world::simulate_wind()
{
	WIND.calculate_wind(world_time, world_width, world_height);
}

Vector world::get_wind(POINT pt)
{	
	if (WIND.inBox(pt))
	{
		Vector out;
		out = WIND.dir;
		out.MultiplyVectorByScalar(WIND.magnitude);
		return out;
	}
	else
	{
		return Vector(0, 0, 0);
	}
}

void world::simulate_rain()
{
	RAIN.calculate_rain(world_time, world_width, world_height);
	if (find_spore)
	{
		

	}
}

void world::simulate_sun()
{
	SUN->calculate_sun(world_time, world_width);
}

void world::simulate_land()
{
	//Nature boxes should be simulated before land, so that effects of nature on land can be calculated.



	//Assumes that rain has already been calculated, leaving only changing moisture if landplot intersects with rainbox.

	for (int i = 0; i < land_plots.size(); i++)
	{
		//Temperature
		land_plots[i].calculate_temperature(world_time);


		//Moisture
		if (isOverlapping(RAIN.brc, land_plots[i].brc))
		{
			//It's raining in this landplot.
			//Increase moisture in landplot according to the strength (i.e. magnitude) of the rain.
			land_plots[i].change_moisture(RAIN.magnitude);
		}

		for (int j = 0; j < rains.size(); j++)
		{
			if (isOverlapping(rains[j].brc, land_plots[i].brc))
			{
				//It's raining in this landplot.
				//Increase moisture in landplot according to the strength of the rain.
				land_plots[i].change_moisture(rains[j].magnitude);
			}
		}


		if (land_plots[i].temperature > 70)
		{

			land_plots[i].change_moisture(-.005 * (land_plots[i].temperature / MAX_TEMPERATURE));
		}
		if (land_plots[i].temperature > 95)
		{

			land_plots[i].change_moisture(-1 * (land_plots[i].temperature / MAX_TEMPERATURE));
		}

		//Nutrients
		land_plots[i].change_nutrients(land_plots[i].nut_regen * delta_time);


		//Sunlight
		if (isOverlapping(SUN->brc, land_plots[i].brc))
		{
			//Sun is shining in this landplot.
			//Set sunlight.
			land_plots[i].set_sunlight(SUN->magnitude);
		}
		else
		{
			land_plots[i].set_sunlight(50);
		}



		land_plots[i].calculate_fertility();
	}
}

void world::simulate_world(float step_time)
{
	//step_time is how much time should be simulated since last call to simulate_world
	//delta_time is saved and will be used.
	//reconcile world_time (simulation variable) with running_time (app variable)??

	float final_time = world_time + step_time;


	while (world_time < final_time)
	{
		world_time += delta_time;
		//Think of all the nature boxes as sort of like the forces on the land. 
		//When they're all calculated, the land can be determined and the order of nature box calculations doesn't matter.
		simulate_wind();
		simulate_rain();
		simulate_sun();
		//Think of the land as one of the factors that affect entities (nature boxes do as well)
		simulate_land();


		


		//Now, simulate entities
		DynQueue<view_obj*>::Node* curObj = nullptr;
		objects.traverse(&curObj);
		while (curObj)
		{
			if (curObj->data->alive == false)
			{
				DynQueue<view_obj*>::Node* temp = curObj;
				objects.traverse(&curObj);
				objects.deleteNode(temp);
			}
			else
			{
				curObj->data->accelerate(delta_time);
				curObj->data->view_move(delta_time);
				
				barrier* bar = nullptr;
				if (checkBarrierCollisions(curObj->data->view_loc, &bar))
				{
					curObj->data->Collision(50000, 0, 0);
					if (curObj->data->type == "cell")
					{
						cell* pCell = reinterpret_cast<cell*>(curObj->data);
						pCell->feel_barrier();
					}
				}
				

				if (!PtInRect(&brc, curObj->data->view_loc))
				{
					//World-boundary collision
					curObj->data->Collision(300, 0, 0);
				}

				if (curObj->data->type == "cell")
				{
					int num_objects = objects.getCount();
					cell* pCell = reinterpret_cast<cell*>(curObj->data);
					pCell->see(&objects, num_objects);
					if (pCell->view_loc.x >= world_width || pCell->view_loc.x <= 0 || pCell->view_loc.y >= world_height || pCell->view_loc.y <= 0)
					{
						//pCell->force_stop();
						//pCell->Collision(100, 0, 0);
						pCell->feel_barrier();
					}
					pCell->simulate();
				}
				if (curObj->data->type == "plant")
				{
					plant* pPlant = reinterpret_cast<plant*>(curObj->data);
					pPlant->grow();
				}
				if (curObj->data->type == "spore")
				{
					spore* pSpore = reinterpret_cast<spore*>(curObj->data);

					//Hacky temporary solution. Apply wind.
					//I consider it hacky because, really, wind should apply to all particles.. But I'm worried how that will affect the simulation.

					if (PtInRect(&WIND.brc, pSpore->view_loc))
					{
						Vector force = WIND.dir;
						force = force.MultiplyVectorByScalar(WIND.magnitude);
						pSpore->applyForce(force);
					}
					
					//Should aging be applied to an object externally in world simulation function? Or should aging be part of the object's internal simulation function?
					pSpore->life_time += delta_time;
					pSpore->simulate();
				}
				if (curObj->data->type == "food")
				{
					food* pFood = reinterpret_cast<food*>(curObj->data);
					if (PtInRect(&WIND.brc, pFood->view_loc))
					{
						Vector force = WIND.dir;
						force = force.MultiplyVectorByScalar(WIND.magnitude);
						pFood->applyForce(force);
					}
				}




				objects.traverse(&curObj);
			}
		}




	}


}




food* world::spawn_food(POINT pt, bool attach)
{
	food* fd1 = new food(pt);
	fd1->mass = 20;
	if (attach)
		fd1->attach_to_point(winPoint_to_myPoint(pt), 1000, 25);
	objects.enqueue(reinterpret_cast<view_obj*>(fd1));
	return fd1;
}

void world::spawn_cell(POINT pt, int id)
{
	cell* c1 = new cell(pt, id);
	c1->mass = 50;
	objects.enqueue(reinterpret_cast<view_obj*>(c1));
}

void world::spawn_plant(POINT pt, Vector dir, int seg_length)
{
	plant* p1 = new plant(pt, dir, seg_length, this);
	p1->mass = 100;
	p1->attach_to_point(winPoint_to_myPoint(pt), 1000, 1);
	objects.enqueue(reinterpret_cast<view_obj*>(p1));
}

void world::spawn_spore(POINT pt, Vector dir, float initial_speed)
{
	if (int(dir.get_size()*10) > 10)
	{
		//not a unit vector or a 0 vector.
		return;
	}
	
	//dir may be left as 0, if desired, but it must be either 0 or a unit vector.
	spore* s1 = new spore(pt, this);
	s1->growth_dir = dir;
	s1->mass = 5;
	s1->vel = s1->growth_dir.MultiplyVectorByScalar(initial_speed);
	objects.enqueue(reinterpret_cast<view_obj*>(s1));
}

void world::spawn_linear_barrier(POINT origin, float c0, float c1,float xmin, float xmax)
{
	float* coeff = new float[2];
	coeff[0] = c0;
	coeff[1] = c1;

	barrier* b1 = new barrier(coeff, 1, xmin, xmax, 5, 20, origin);
	b1->mass = 500;
	objects.enqueue(reinterpret_cast<view_obj*>(b1));
	barriers.enqueue(b1);

	POINT tp;
	tp.x = 392;
	tp.y = 700;
	bool test = b1->isCollision(tp);

}

void world::spawn_parabolic_barrier(POINT origin, float c0, float c1, float c2, float xmin, float xmax)
{
	float* coeff = new float[3];
	coeff[0] = c0;
	coeff[1] = c1;
	coeff[2] = c2;

	barrier* b1 = new barrier(coeff, 2, xmin, xmax, 5, 20, origin);
	b1->mass = 500;
	objects.enqueue(reinterpret_cast<view_obj*>(b1));
	barriers.enqueue(b1);
}

void world::spawn_cubic_barrier(POINT origin, float c0, float c1, float c2, float c3, float xmin, float xmax)
{
	float* coeff = new float[4];
	coeff[0] = c0;
	coeff[1] = c1;
	coeff[2] = c2;
	coeff[3] = c3;

	barrier* b1 = new barrier(coeff, 3, xmin, xmax, 5, 20, origin);
	b1->mass = 500;
	objects.enqueue(reinterpret_cast<view_obj*>(b1));
	barriers.enqueue(b1);
}


bool world::checkBarrierCollisions(POINT pt, barrier** bar)
{
	//If at least one barrier collides, returns true;.
	//Saves the address of the first barrier that collides in bar parameter.


	DynQueue<barrier*>::Node* curObj = nullptr;
	barriers.traverse(&curObj);
	while (curObj)
	{
		if (curObj->data->alive)
		{
			if (curObj->data->isCollision(pt))
			{
				*bar = curObj->data;
				return true;
			}
		}
		barriers.traverse(&curObj);
	}
	*bar = nullptr;
	return false;
}

bool world::isOverlapping(RECT r1, RECT r2)
{
	bool p = (r1.left <= r2.left && r2.left <= r1.right);
	bool q = (r1.top <= r2.bottom && r2.bottom <= r1.bottom);
	bool r = (r1.top <= r2.top && r2.top <= r1.bottom);
	bool s = (r1.left <= r2.right && r2.right <= r1.right);

	bool vert = q || r;

	return vert && (p || s);
}

land* world::getLand(POINT pt)
{
//	for (int i = 0; i < land_plots.size(); i++)
//	{
//		if (land_plots[i].point_in_land(pt))
//		{
//			return &land_plots[i];
//		}
//	}

	//check that pt is in world.
	if (PtInRect(&brc, pt))
	{
		int i = pt.x / land_width;
		int j = pt.y / land_height;

		if (i == num_columns)
		{
			//Accounting for the remainder of world_width which may have been included in the border plots.
			i = num_columns - 1;
		}
		if (j == num_rows)
		{
			//Accounting for the remainder of world_height which may have been included in the border plots.
			j = num_rows - 1;
		}
		int idx = i * num_rows + j;
		return &land_plots[idx];

	}
	return nullptr;
}